lib/Parse/ParseExpr.cpp

```diff
@@ -3127,19 +3143,88 @@ ParserStatus Parser::parseExprList(tok leftTok, tok rightTok,
     return status;
 
   // Parse the closure.
-  ParserResult<Expr> closure =
-    parseTrailingClosure(SourceRange(leftLoc, rightLoc));
-  status |= closure;
-  if (closure.isNull())
-    return status;
+  status |=
+      parseTrailingClosures(isExprBasic, SourceRange(leftLoc, rightLoc),
+                            trailingLBrace, trailingRBrace, trailingClosures);
+  return status;
+}
 
-  // Record the trailing closure.
-  trailingClosure = closure.get();
+/// Check whether upcoming trailing closure actually represents a block of
+/// labeled trailing closures e.g.
+/// foo() {
+///  <label-1>: { ... }
+///  ...
+/// }
+static bool isBlockOfMultipleTrailingClosures(bool isExprBasic, Parser &P) {
+  assert(isValidTrailingClosure(isExprBasic, P));
+  Parser::BacktrackingScope backtrack(P);
+  P.consumeToken(tok::l_brace);
 
-  return status;
+  if (!P.Tok.canBeArgumentLabel() || P.Tok.is(tok::kw__))
+    return false;
+
+  auto text = P.Tok.getText();
+  if (text[0] == '$')
+    return false;
```

Comment:
You could this before starting the backtracking.
```cpp
if (!P.peekToken().canBeArgumentLabel() || P.peekToken().is(tok::kw__) ||
    P.peekToken().getText()[0] == '$')
  return false;

Parser::BacktrackingScope backtrack(P);
P.consumeToken(tok::l_brace);
P.consumeToken();
```

Reply:
I can’t do that because its labeled blocks enclosed in braces, so I need to consume or peek through the left brace first somehow first...

Reply:
Actually I take that back, I could do that to avoid backtracking

Reply:
But `peekToken()` can look through the brace..

Reply:
Fixed!